static/app/views/performance/newTraceDetails/traceApi/useTraceTree.tsx

```diff
@@ -0,0 +1,121 @@
+import {useEffect, useMemo, useReducer, useRef, useState} from 'react';
+
+import type {TraceSplitResults} from 'sentry/utils/performance/quickTrace/types';
+import type {QueryStatus, UseApiQueryResult} from 'sentry/utils/queryClient';
+import useApi from 'sentry/utils/useApi';
+import useOrganization from 'sentry/utils/useOrganization';
+import useProjects from 'sentry/utils/useProjects';
+import type {ReplayRecord} from 'sentry/views/replays/types';
+
+import {TraceTree} from '../traceModels/traceTree';
+
+import type {TraceMetaQueryResults} from './useTraceMeta';
+
+type UseTraceTreeParams = {
+  metaResults: TraceMetaQueryResults;
+  replayRecord: ReplayRecord | null;
+  traceResults: UseApiQueryResult<
+    TraceSplitResults<TraceTree.Transaction> | undefined,
+    any
+  >;
+  traceSlug?: string;
+};
+
+function getTraceViewQueryStatus(
+  traceQueryStatus: QueryStatus,
+  traceMetaQueryStatus: QueryStatus
+): QueryStatus {
+  if (traceQueryStatus === 'error' || traceMetaQueryStatus === 'error') {
+    return 'error';
+  }
+
+  if (traceQueryStatus === 'pending' || traceMetaQueryStatus === 'pending') {
+    return 'pending';
+  }
+
+  return 'success';
+}
+
+export function useTraceTree({
+  traceResults,
+  metaResults,
+  traceSlug,
+  replayRecord,
+}: UseTraceTreeParams): TraceTree {
+  const api = useApi();
+  const {projects} = useProjects();
+  const organization = useOrganization();
+
+  const [tree, setTree] = useState<TraceTree>(TraceTree.Empty());
+  const loadingTraceRef = useRef<TraceTree | null>(null);
+  const [_, rerender] = useReducer(x => (x + 1) % Number.MAX_SAFE_INTEGER, 0);
```

Comment:
I'm generally in favor of this, but this rerender will cause issues. Note that even though this hook updates, the tree reference stays the same, meaning if we pass the tree to child components they risk never updating. This used to work because it lived inside the view component, but abstracting it away will break this in silent ways